### PR TITLE
fix config path & format

### LINKED ISSUE #13: Respect old DVC user-id
```python
import json
from shutil import copy
from pathlib import Path
from appdirs import user_config_dir

uid = generate_id() # see above

# DVC backwards-compatibility
old = Path(user_config_dir("dvc/user_id", "iterative"))
# cross-product path
new = Path(user_config_dir("iterative/telemetry", False))

if old.exists() and not new.exists():
    new.write_text(uid := json.load(old.open())["user_id"])

# only for non-DVC packages:
if not old.exists():
    # write legacy file in case DVC is installed later
    old.write_text(f'{{"user_id": "{uid}"}}')
```

### PR DESCRIPTION
**Pull request by casperdcl**
- [x] for cross-platform compatibility: use `appdirs.user_config_dir`'s `appauthor` even on non-Windows systems
- [x] look for legacy DVC id (fixes #13, replaces/closes #31)
- [x] ~write to legacy file~
- [x] ~expose `write_legacy` for DVC to set to `False`~
- [x] ~do not use JSON~ use JSON

vis:

- spec https://www.notion.so/iterative/Telemetry-Specification-1a0197b8e34d43a89777a4c9a00ad1fe#f24a474b773b4fcd8ff65fc65044343d
- implementation: https://github.com/iterative/terraform-provider-iterative/blob/7c800a9d178efe11ade71248915166d6b651453c/iterative/utils/analytics.go#L210

### CODE REVIEW

File: docs/gen_ref_pages.py [hunk 1]
```diff
@@ -18,8 +18,8 @@
     elif parts[-1] == "__main__":
         continue
 
-    with mkdocs_gen_files.open(full_doc_path, "w") as fd:
+    with mkdocs_gen_files.open(full_doc_path, "w") as fobj:
         identifier = ".".join(parts)
-        print("::: " + identifier, file=fd)
+        print("::: " + identifier, file=fobj)
```
[the remark this answers is not in the document]
Comment (casperdcl) [reply docs/gen_ref_pages.py:23]: I also prefer `f` but unfortunately linters usually discourage it. `file` clashes with a [deprected builtin](https://docs.python.org/2.7/library/functions.html#file) so also often causes linters to complain.

File: src/iterative_telemetry/__init__.py [hunk 2]
```diff
@@ -187,39 +187,55 @@ def _system_info():
     raise NotImplementedError
 
 
+def generate_id():
```
[the remark this answers is not in the document]
Comment (casperdcl) [reply src/iterative_telemetry/__init__.py:190]: changed

File: src/iterative_telemetry/__init__.py [hunk 3]
```diff
@@ -187,39 +187,55 @@ def _system_info():
...
-                with open(fname, "w", encoding="utf8") as fobj:
+            user_id = read_user_id(config_file)
+            if user_id is None:
+                if config_file_old.parent.is_dir():
+                    with FileLock(lockfile_old, timeout=5):
+                        user_id = read_user_id(config_file_old)
+                    if user_id is None:
+                        user_id = generate_id()
+                with config_file.open(mode="w", encoding="utf8") as fobj:
                     json.dump({"user_id": user_id}, fobj)
```
[the remark this answers is not in the document]
Comment (casperdcl) [reply src/iterative_telemetry/__init__.py:226]: fixed

File: src/iterative_telemetry/__init__.py [hunk 4]
```diff
@@ -187,39 +187,55 @@ def _system_info():
...
+    config_file = Path(
+        user_config_dir(os.path.join("iterative", "telemetry"), False)
+    )
+    config_file.parent.mkdir(mode=0o755, parents=True, exist_ok=True)
+    lockfile = str(config_file.with_suffix(".lock"))
+    # DVC backwards-compatibility
+    config_file_old = Path(
+        user_config_dir(os.path.join("dvc", "user_id"), "iterative")
+    )
+    lockfile_old = str(config_file_old.with_suffix(".lock"))
```
[the remark this answers is not in the document]
Comment (casperdcl) [reply src/iterative_telemetry/__init__.py]: I also initially felt like doing the careful checking & moving, but... I suppose it's not exactly critical. Cleaner code is also nice.

File: src/iterative_telemetry/__init__.py [hunk 5]
```diff
@@ -187,39 +187,60 @@ def _system_info():
...
+            return json.load(fobj)["user_id"]
+    except (FileNotFoundError, ValueError, KeyError):
+        pass
+    return None
+
+
+def _read_user_id_locked(config_file: Path):
+    lockfile = str(config_file.with_suffix(".lock"))
+    if config_file.parent.is_dir():
+        with FileLock(lockfile, timeout=5):
```
Comment (efiop) [inline src/iterative_telemetry/__init__.py:207]: Timeout exception from here will fall through and result in an error that will say that it failed to acquire new lockfile when in fact this is about an old lockfile.

Comment (casperdcl) [reply src/iterative_telemetry/__init__.py:207]: indeed, the "fall through and result in an error (though really a debug message) that will say that it failed to acquire" is intentional, but the "new lockfile when in fact this is about an old lockfile" is indeed misleading. Fixed now :)

File: src/iterative_telemetry/__init__.py [hunk 6]
```diff
@@ -187,39 +187,64 @@ def _system_info():
...
+            user_id = _read_user_id(config_file)
+            if user_id is None:
+                try:
+                    user_id = _read_user_id_locked(config_file_old)
+                except Timeout:
+                    logger.debug(
+                        "Failed to acquire %s",
+                        config_file_old.with_suffix(".lock"),
+                    )
+                    return None
```
Comment (efiop) [inline src/iterative_telemetry/__init__.py:243]: Let's keep this simple and readable and move the try&except _read_user_id_locked(

```suggestion
                user_id = _read_user_id_locked(config_file_old)
```

**Pull request merged**